src/algebraic_geometry/EllipticCurve/weierstrass.lean

```diff
@@ -0,0 +1,287 @@
+@[simps] def base_change : weierstrass_curve A :=
+⟨algebra_map R A C.a₁, algebra_map R A C.a₂, algebra_map R A C.a₃, algebra_map R A C.a₄,
+algebra_map R A C.a₆⟩
+
+@[simp] lemma base_change_b₂ : (C.base_change A).b₂ = algebra_map R A C.b₂ :=
+by { simp only [b₂, base_change_a₁, base_change_a₂], map_simp }
+
+@[simp] lemma base_change_b₄ : (C.base_change A).b₄ = algebra_map R A C.b₄ :=
+by { simp only [b₄, base_change_a₁, base_change_a₃, base_change_a₄], map_simp }
+
+@[simp] lemma base_change_b₆ : (C.base_change A).b₆ = algebra_map R A C.b₆ :=
+by { simp only [b₆, base_change_a₃, base_change_a₆], map_simp }
+
+@[simp] lemma base_change_b₈ : (C.base_change A).b₈ = algebra_map R A C.b₈ :=
+by { simp only [b₈, base_change_a₁, base_change_a₂, base_change_a₃, base_change_a₄, base_change_a₆],
+     map_simp }
+
+@[simp] lemma base_change_c₄ : (C.base_change A).c₄ = algebra_map R A C.c₄ :=
+by { simp only [c₄, base_change_b₂, base_change_b₄], map_simp }
+
+@[simp] lemma base_change_c₆ : (C.base_change A).c₆ = algebra_map R A C.c₆ :=
+by { simp only [c₆, base_change_b₂, base_change_b₄, base_change_b₆], map_simp }
+
+@[simp] lemma base_change_Δ : (C.base_change A).Δ = algebra_map R A C.Δ :=
+by { simp only [Δ, base_change_b₂, base_change_b₄, base_change_b₆, base_change_b₈], map_simp }
+
+end base_change
+
+section torsion_polynomial
+
+/-! ### 2-torsion polynomials -/
+
+/-- A cubic polynomial whose discriminant is a multiple of the Weierstrass curve discriminant.
+If `C` is an elliptic curve over a field `R` of characteristic different from 2, then its roots over
+a splitting field of `R` are precisely the x-coordinates of the non-zero 2-torsion points of `C`. -/
+def two_torsion_polynomial : cubic R := ⟨4, C.b₂, 2 * C.b₄, C.b₆⟩
+
+lemma two_torsion_polynomial_disc : C.two_torsion_polynomial.disc = 16 * C.Δ :=
+by { dsimp [two_torsion_polynomial, cubic.disc], ring1 }
+
+end torsion_polynomial
+
+end weierstrass_curve
+
+/-! ## Elliptic curves -/
+
+/-- An elliptic curve over a commutative ring. Note that this definition is only mathematically
+accurate for certain rings whose Picard group has trivial 12-torsion, such as a field or a PID. -/
+@[ext] structure elliptic_curve (R : Type u) [comm_ring R] extends weierstrass_curve R :=
+(Δ' : Rˣ) (coe_Δ' : ↑Δ' = to_weierstrass_curve.Δ)
+
+instance : inhabited $ elliptic_curve ℚ :=
+⟨⟨⟨0, 0, 1, -1, 0⟩, ⟨37, 37⁻¹, by norm_num1, by norm_num1⟩, by { dsimp, ring1 }⟩⟩
+
+namespace elliptic_curve
+
+variables [comm_ring R] (E : elliptic_curve R)
+
+/-- The j-invariant `j` of an elliptic curve, which is invariant under isomorphisms over `R`. -/
+@[simp] def j : R := ↑E.Δ'⁻¹ * E.c₄ ^ 3
+
+lemma two_torsion_polynomial.disc_ne_zero [nontrivial R] [invertible (2 : R)] :
+  E.two_torsion_polynomial.disc ≠ 0 :=
+λ hdisc, E.Δ'.ne_zero $ (is_unit_of_invertible $ 2 ^ 4).mul_left_cancel $
+by linear_combination hdisc - E.two_torsion_polynomial_disc
+   with { normalization_tactic := `[rw [coe_Δ'], ring1] }
```

Comment:
I suggest that you add
```
lemma two_torsion_polynomial.is_unit_disc_iff_Δ [invertible (2 : R)] :
  is_unit (C.two_torsion_polynomial.disc) ↔ is_unit C.Δ :=
begin
  rw two_torsion_polynomial_disc,
  refine ⟨is_unit_of_mul_is_unit_right, λ h, _⟩,
  convert (is_unit_of_invertible $ 2 ^ 4).mul h,
  ring,
end
```
and golf this as
```suggestion
((E^.two_torsion_polynomial.is_unit_disc_iff_Δ).2 $ E.coe_Δ' ▸ E.Δ'.is_unit).ne_zero
```
with a new lemma

Reply:
Thanks, but what does `E^` do exactly?

Reply:
If you have a declaration `foo1.bar1.bar2` then `C.bar1.bar2` doesn't work (it would be interpreted as `foo2.bar2 (foo1.bar1 C)`, but `C^.bar1.bar2` works.

Reply:
Is writing the notation `^.`  just a hack for this or is there an explanation in the manual somewhere?

Reply:
It's some special syntax I learned from reading the code (e.g. [here](https://github.com/leanprover-community/mathlib/blob/cc3c8aa55d1829f1fe5f53f9849ba859be355f7d/src/ring_theory/ideal/operations.lean#L1823)). I don't know where (or if) it's documented.